adsmp/tasks.py

```diff
@@ -73,58 +76,100 @@ def task_route_record(bibcode, force=False, delayed=1):
-            # build the record and send it to solr
-            solr_doc = solr_updater.transform_json_record(r)
-            solr_updater.update_solr(solr_doc, app.conf.get('SOLR_URLS'))
-            app.update_processed_timestamp(bibcode)
+    for bibcode in bibcodes:
+        r = app.get_record(bibcode)
+    
+        if r is None:
+            logger.error('The bibcode %s doesn\'t exist!', bibcode)
+            continue
+    
+        bib_data_updated = r.get('bib_data_updated', None)
+        orcid_claims_updated = r.get('orcid_claims_updated', None)
+        nonbib_data_updated = r.get('nonbib_data_updated', None)
+        fulltext_updated = r.get('fulltext_updated', None)
+        metrics_updated = r.get('metrics_updated', None)
+    
+        year_zero = '1972'
+        processed = r.get('processed', adsputils.get_date(year_zero))
+        if processed is None:
+            # It was never sent to Solr
+            processed = adsputils.get_date(year_zero)
+    
+        is_complete = all([bib_data_updated, orcid_claims_updated, nonbib_data_updated])
+    
+        if is_complete:
+            if force is False and all([bib_data_updated and bib_data_updated < processed,
+                   orcid_claims_updated and orcid_claims_updated < processed,
+                   nonbib_data_updated and nonbib_data_updated < processed]):
+                logger.debug('Nothing to do for %s, it was already indexed/processed', bibcode)
+                continue
+            else:
+                # build the solr record
+                if update_solr:
+                    batch.append(solr_updater.transform_json_record(r))
+                # get data for metrics
+                if update_metrics:
+                    m = r.get('metrics', None)
+                    if m:
+                        if r.get('processed'):
+                            batch_update.append((bibcode, m))
+                        else:
+                            batch_insert((bibcode, m))
         else:
-            
-            logger.warn('{bibcode} is missing bib data, even with force=True, this cannot proceed'.format(
-                            bibcode=bibcode))
+            # if forced and we have at least the bib data, index it
+            if force is True and bib_data_updated:
+                logger.warn('Forced indexing of: %s (metadata=%s, orcid=%s, nonbib=%s, fulltext=%s, metrics=%s)' % \
+                            (bibcode, bib_data_updated, orcid_claims_updated, nonbib_data_updated, fulltext_updated, \
+                             metrics_updated))
+                # build the record and send it to solr
+                if update_solr:
+                    batch.append(solr_updater.transform_json_record(r))
+                # get data for metrics
+                if update_metrics:
+                    m = r.get('metrics', None)
+                    if m:
+                        if r.get('processed'):
```

Comment:
I assume processed tells us if metrics has updated this bibcode before.  I don' think it is definitive on whether we should perform a sql insert or update on the Metrics database.  Perhaps the Metrics database was lost or maybe we updated the Metrics database but failed to set the processed flag in our local database.  Can we rely on the processed flag and the metrics database to stay in sync?  
Sorry if comment appears twice, I thought I created it but then could not find it.